File: cpp/stacks_queues/queues.cc
#include <iostream>

using namespace std;

// Implement a Queue using a linked list
class Queue {
  private:
  // Node class definition
    class Node {
      public:
        int data;
        Node* next;

      Node(int data, Node* next) {
        this->data = data;
        this->next = next;
      }
    };

  Node* head;
  Node* tail;

  public:
    Queue() {
      this->head = NULL;
      this->tail = NULL;
    }

    // insert from the tail
    void insert(int item) {
      Node* new_node = new Node(item, NULL);
      if (this->tail != NULL) {
        this->tail->next = new_node;
      }
      this->tail = new_node;
      if (this->head == NULL) {
        this->head = this->tail;
      }
    }

    // remove from the head
    int remove() {
      if (this->head == NULL) {
        throw "Queue Empty!";
      }
      int ret_val = this->head->data;
      Node* temp_head = this->head;
      this->head = this->head->next;
      if (this->head == NULL) {
        this->tail = NULL;
      }
      free(temp_head);
      return ret_val;
    }

    int peek() {
      if (this->head == NULL) {
        throw "Queue Empty!";
      }
      return (this->head)->data;
    }

    bool isEmpty() {
      return (this->head == NULL);
    }

};

int main() {
  Queue q;
  q.insert(1);
  cout << (q.peek()) << endl;
  q.insert(2);
  cout << (q.peek()) << endl;
  q.insert(3);
  cout << (q.peek()) << endl;
  cout << (q.isEmpty()) << endl;

  cout << (q.remove()) << endl;
  cout << (q.remove()) << endl;
  cout << (q.remove()) << endl;
  cout << (q.isEmpty()) << endl;
}

File: python/ctci/stacks_and_queues/queues.py
# python implementation of a queue with a linked list (ADT)
class Queue(object):

    class Node(object):
        def __init__(self, data=None, next=None):
            self.data = data
            self.next = next

    def __init__(self):
        self.first = None
        self.last = None

    def add(self, item):
        # First item case. Both should point to the same object
        new_node = self.Node(item, None)
        if self.first is None and self.last is None:
            self.first, self.last = new_node, new_node
        else:
            self.last.next = new_node
            self.last = new_node

    def remove(self):
        if self.first is None:
            raise Exception('Queue Empty')

        # When first and last meet
        ret_val = self.first.data
        if self.first is self.last:
            self.first, self.last = None, None
        else:
            self.first = self.first.next
        return ret_val


    def peek(self):
        if self.first is None:
            raise Exception('Queue Empty')
        return self.first.data

    def isEmpty(self):
        if self.first is None:
            return True
        return False

if __name__ == '__main__':
    q = Queue()
    q.add(1)
    print(q.peek())
    q.add(2)
    q.add(3)
    print(q.peek())
    print(q.isEmpty())
    print(q.remove())
    print(q.remove())
    print(q.remove())
    print(q.isEmpty())
